File: 백준/Gold/14852. 타일 채우기 3/타일 채우기 3.py
MOD = 1000000007

def tile_fill(n):
    if n == 1:
        return 2
    elif n == 2:
        return 7

    dp = [0] * (n + 1)
    dp_sum = [0] * (n + 1)

    dp[1] = 2
    dp[2] = 7
    dp_sum[2] = 1  # dp[0]의 값은 1로 간주

    for i in range(3, n + 1):
        dp_sum[i] = (dp_sum[i - 1] + dp[i - 3]) % MOD
        dp[i] = (2 * dp[i - 1] + 3 * dp[i - 2] + 2 * dp_sum[i]) % MOD

    return dp[n]

# 예시 실행
n = int(input())
print(tile_fill(n))

File: 백준/Gold/14852. 타일 채우기 3/타일 채우기 3.java
import java.util.*;
import java.io.*;

public class Main {

    public static void main(String[] args) throws IOException{
    	BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    	N = Integer.parseInt(br.readLine());
    	dpSum = new long[N+1];
    	dp = new long[N+1];
    	
    	dp[1] = 2;
    	if( N >= 2)
    		dp[2] = 7;
    	
    	dpSum[0] = 1;
    	
    	for(int i = 3; i<=N;i++) {
    		dp[i] = (2*dp[i-1] + 3 * dp[i-2] + 2 * dpSum[i-3]) % MOD;
    		dpSum[i-2] = (dpSum[i-3] + dp[i-2]) % MOD;
    	} 
    	System.out.println(dp[N]);
    }
    static int N;
    static long dp[], dpSum[];
    static int MOD = (int)1e9+7;
}